File: scripts/localize.mjs
#!/usr/bin/env node
/**
 * RF4 游戏汉化工具 - 命令行版本
 * 
 * 用法：
 *   node scripts/localize.mjs <俄服文件路径> [输出文件路径]
 * 
 * 示例：
 *   node scripts/localize.mjs "C:\Users\book\Desktop\汉化\俄服文件\resources.assets"
 *   node scripts/localize.mjs resources.assets resources_cn.assets
 */

import { readFileSync, writeFileSync } from 'fs'
import { basename, dirname, join } from 'path'
import { fileURLToPath } from 'url'

const __dirname = dirname(fileURLToPath(import.meta.url))

// 解析命令行参数
const inputFile = process.argv[2]
const outputFile = process.argv[3] || inputFile.replace('.assets', '_cn.assets')

if (!inputFile) {
  console.error('用法: node scripts/localize.mjs <输入文件> [输出文件]')
  process.exit(1)
}

console.log('🎮 RF4 游戏汉化工具')
console.log(`📂 输入文件: ${inputFile}`)
console.log(`📤 输出文件: ${outputFile}`)
console.log('')

// 读取文件
console.log('⏳ 正在读取文件...')
const buffer = readFileSync(inputFile)
console.log(`✅ 文件大小: ${(buffer.length / (1024 * 1024)).toFixed(2)} MB`)

// 提取 UTF-8 文本
console.log('⏳ 正在提取文本...')
const content = buffer.toString('utf-8')

// 查找俄语文本并翻译（简化版：使用预设映射表）
console.log('⏳ 正在翻译文本...')

// 这里应该调用翻译 API，但由于需要 API Key，我们提供一个示例映射
// 实际使用时，用户需要：
// 1. 注册百度翻译 API（https://fanyi-api.baidu.com/）
// 2. 获取 APP ID 和密钥
// 3. 替换下面的 translateText 函数

const translationMap = new Map()

function translateText(text) {
  // TODO: 实现真实的翻译逻辑
  // 示例：调用百度翻译 API
  /*
  const appId = 'YOUR_APP_ID'
  const key = 'YOUR_KEY'
  const salt = Date.now()
  const sign = md5(appId + text + salt + key)
  
  const response = await fetch(
    `https://fanyi-api.baidu.com/api/trans/vip/translate?q=${encodeURIComponent(text)}&from=ru&to=zh&appid=${appId}&salt=${salt}&sign=${sign}`
  )
  const data = await response.json()
  return data.trans_result[0].dst
  */
  
  // 当前返回原文（占位符）
  return `[待翻译]${text}`
}

// 查找所有俄语文本段
const russianRegex = /[\u0400-\u04FF]{4,}/g
let matches
let translatedCount = 0
let modifiedContent = content

while ((matches = russianRegex.exec(content)) !== null) {
  const originalText = matches[0]
  
  // 跳过已翻译的或特殊标记
  if (originalText.startsWith('[待翻译]')) continue
  
  // 翻译文本
  const translated = translateText(originalText)
  
  // 替换（注意：由于字符串长度变化，直接替换会导致偏移错误）
  // 完整实现需要重新构建二进制结构
  translatedCount++
  
  if (translatedCount % 100 === 0) {
    console.log(`  已处理 ${translatedCount} 段文本...`)
  }
}

console.log(`✅ 共找到 ${translatedCount} 段俄语文本`)

// 写入文件（当前为占位符，实际需要正确重建 Unity 序列化格式）
console.log('⚠️  警告：当前版本仅为演示框架，未实现完整的 Unity 序列化文件重建逻辑')
console.log('💡 建议使用专业工具如 UnityPy 或 AssetStudio 进行处理')

// writeFileSync(outputFile, Buffer.from(modifiedContent, 'utf-8'))
// console.log(`✅ 汉化完成！输出文件: ${outputFile}`)

console.log('\n📖 完整实现方案：')
console.log('1. 安装 UnityPy: pip install UnityPy')
console.log('2. 使用 Python 脚本解析和重建 .assets 文件')
console.log('3. 或参考提供的样本文件进行手动对照翻译')

File: scripts/localize.py
#!/usr/bin/env python3
"""
RF4 游戏汉化工具 - Python 版本（推荐）

依赖安装：
  pip install UnityPy requests

用法：
  python scripts/localize.py <俄服文件路径> [输出文件路径]

示例：
  python scripts/localize.py "C:\Users\book\Desktop\汉化\俄服文件\resources.assets"
  python scripts/localize.py resources.assets resources_cn.assets
"""

import sys
import os
import re
import json
import time
from pathlib import Path

try:
    import UnityPy
except ImportError:
    print("❌ 缺少依赖库 UnityPy")
    print("请运行: pip install UnityPy")
    sys.exit(1)

try:
    import requests
except ImportError:
    print("❌ 缺少依赖库 requests")
    print("请运行: pip install requests")
    sys.exit(1)


class RF4Localizer:
    """RF4 游戏汉化器"""
    
    def __init__(self, baidu_app_id=None, baidu_key=None):
        self.baidu_app_id = baidu_app_id or os.getenv('BAIDU_APP_ID')
        self.baidu_key = baidu_key or os.getenv('BAIDU_KEY')
        self.translated_count = 0
        self.cache = {}  # 翻译缓存
        
    def translate_with_baidu(self, text, from_lang='ru', to_lang='zh'):
        """使用百度翻译 API"""
        if not self.baidu_app_id or not self.baidu_key:
            print(f"⚠️  未配置百度翻译 API，跳过翻译: {text[:50]}...")
            return text
        
        # 检查缓存
        if text in self.cache:
            return self.cache[text]
        
        import hashlib
        import random
        
        salt = str(random.randint(32768, 65536))
        sign_str = f"{self.baidu_app_id}{text}{salt}{self.baidu_key}"
        sign = hashlib.md5(sign_str.encode('utf-8')).hexdigest()
        
        url = "https://fanyi-api.baidu.com/api/trans/vip/translate"
        params = {
            'q': text,
            'from': from_lang,
            'to': to_lang,
            'appid': self.baidu_app_id,
            'salt': salt,
            'sign': sign
        }
        
        try:
            response = requests.get(url, params=params, timeout=10)
            data = response.json()
            
            if 'trans_result' in data and len(data['trans_result']) > 0:
                result = data['trans_result'][0]['dst']
                self.cache[text] = result
                return result
            else:
                print(f"⚠️  翻译失败: {data}")
                return text
        except Exception as e:
            print(f"⚠️  翻译请求异常: {e}")
            return text
    
    def extract_text_from_assets(self, assets_path):
        """从 .assets 文件中提取所有文本"""
        print(f"📂 正在读取: {assets_path}")
        
        env = UnityPy.load(assets_path)
        texts = []
        
        for obj in env.objects:
            if obj.type.name in ["TextAsset", "MonoBehaviour"]:
                try:
                    data = obj.read()
                    if hasattr(data, 'm_Name'):
                        texts.append({
                            'path_id': obj.path_id,
                            'type': obj.type.name,
                            'name': data.m_Name,
                            'data': data
                        })
                except Exception as e:
                    print(f"⚠️  读取对象失败: {e}")
        
        print(f"✅ 找到 {len(texts)} 个文本对象")
        return env, texts
    
    def localize_assets(self, input_path, output_path):
        """执行汉化"""
        start_time = time.time()
        
        # 加载文件
        env, texts = self.extract_text_from_assets(input_path)
        
        # 翻译每个文本对象
        for i, item in enumerate(texts):
            data = item['data']
            
            # 处理 TextAsset
            if item['type'] == 'TextAsset' and hasattr(data, 'm_Script'):
                original_text = data.m_Script
                if original_text and self._contains_russian(original_text):
                    translated = self.translate_with_baidu(original_text)
                    data.m_Script = translated
                    self.translated_count += 1
            
            # 处理 MonoBehaviour（可能包含 name/description 字段）
            elif item['type'] == 'MonoBehaviour':
                if hasattr(data, 'name') and data.name:
                    if self._contains_russian(data.name):
                        data.name = self.translate_with_baidu(data.name)
                        self.translated_count += 1
                
                if hasattr(data, 'description') and data.description:
                    if self._contains_russian(data.description):
                        data.description = self.translate_with_baidu(data.description)
                        self.translated_count += 1
            
            # 进度显示
            if (i + 1) % 100 == 0:
                elapsed = time.time() - start_time
                print(f"  进度: {i+1}/{len(texts)} | 已翻译: {self.translated_count} | 耗时: {elapsed:.1f}s")
        
        # 保存文件
        print(f"\n💾 正在保存: {output_path}")
        with open(output_path, 'wb') as f:
            env.save(f)
        
        elapsed = time.time() - start_time
        print(f"\n✅ 汉化完成！")
        print(f"   翻译文本数: {self.translated_count}")
        print(f"   总耗时: {elapsed:.1f}秒")
        print(f"   输出文件: {output_path}")
    
    def _contains_russian(self, text):
        """检查文本是否包含俄语字符"""
        return bool(re.search(r'[\u0400-\u04FF]', text))


def main():
    if len(sys.argv) < 2:
        print(__doc__)
        sys.exit(1)
    
    input_file = sys.argv[1]
    output_file = sys.argv[2] if len(sys.argv) > 2 else input_file.replace('.assets', '_cn.assets')
    
    if not os.path.exists(input_file):
        print(f"❌ 文件不存在: {input_file}")
        sys.exit(1)
    
    print("🎮 RF4 游戏汉化工具 (Python 版)")
    print("=" * 60)
    
    # 创建汉化器
    localizer = RF4Localizer()
    
    # 执行汉化
    localizer.localize_assets(input_file, output_file)


if __name__ == '__main__':
    main()
